retrieval/scrape.go

```diff
@@ -482,30 +498,92 @@ func (sl *scrapeLoop) run(interval, timeout time.Duration, errc chan<- error) {
 		}
 
 		err := sl.scraper.scrape(scrapeCtx, buf)
-		if err == nil {
-			b := buf.Bytes()
+		if err != nil && errc != nil {
+			errc <- err
+		}
 
-			if total, added, err = sl.append(b, start); err != nil {
-				log.With("err", err).Error("append failed")
+		var b []byte
+		if err == nil {
+			b = buf.Bytes()
+		}
+		// A failed scrape is the same as an empty scrape,
+		// we still call sl.append to trigger stale markers.
+		if total, added, err = sl.append(b, start); err != nil {
+			log.With("err", err).Error("append failed")
+			// The append failed, probably due to a parse error.
+			// Call sl.append again with an empty scrape to trigger stale markers.
+			if _, _, err = sl.append([]byte{}, start); err != nil {
```

Comment:
What way were you thinking of splitting them? Parsing and ingesting are fairly tightly bound as-is.